### PR TITLE
Refactor `distinct` using `static_map` `insert_or_apply`

### PR DESCRIPTION
**Pull request by srinivasyadav18**
## Description
This PR refactors `distinct` using `static_map::insert_or_apply` for **keep_first**, **keep_last** and **keep_none** options.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/stream_compaction/distinct_helpers.hpp
Comment (bdice) [inline cpp/src/stream_compaction/distinct_helpers.hpp:1]: This header is now empty. Can we delete it and remove it from `#include`s?

File: cpp/src/stream_compaction/distinct_helpers.cu
Comment (bdice) [inline cpp/src/stream_compaction/distinct_helpers.cu:1]: Likewise, this file is now empty and should be removed and deleted from CMakeLists.txt.

File: cpp/src/stream_compaction/distinct.cu [hunk 1]
```diff
@@ -32,6 +32,9 @@
 #include <rmm/mr/device/per_device_resource.hpp>
 #include <rmm/resource_ref.hpp>
 
+#include <cuco/static_map.cuh>
```
Comment (PointKernel) [inline cpp/src/stream_compaction/distinct.cu:35]:
```suggestion
#include <cuco/static_map.cuh>
#include <cuco/static_set.cuh>
```

File: cpp/src/stream_compaction/distinct.cu [hunk 2]
```diff
@@ -72,8 +75,146 @@ rmm::device_uvector<cudf::size_type> dipatch_row_equal(
...
+     keys           = keys.begin(),
+     output_indices = output_indices.begin()] __device__(size_type const idx) mutable {
+      if (values[idx] == size_type{1}) {
+        output_indices[idx] = keys[idx];
+      } else {
+        output_indices[idx] = -1;
+      }
+    });
+
+  auto const map_end = thrust::copy_if(
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/stream_compaction/distinct.cu:161]: Good catch

**Comment by srinivasyadav18** [comment]
@bdice @PointKernel, I pushed all the cleanups and minor optimizations. The PR is ready for review.

Also, I don't think we might need extra optimization on **keep_none**. As @bdice mentioned, **transform counting iterator** with `copy_if` should be efficient enough.

File: cpp/src/stream_compaction/distinct.cu [hunk 3]
```diff
@@ -72,8 +78,114 @@ rmm::device_uvector<cudf::size_type> dispatch_row_equal(
...
+    }
+
+    auto const [_, output_end] =
+      map.retrieve_all(thrust::make_discard_iterator(), output_indices.begin(), stream.value());
+    output_indices.resize(thrust::distance(output_indices.begin(), output_end), stream);
+    return output_indices;
+  }
+
+  auto keys   = rmm::device_uvector<size_type>(num_rows, stream, mr);
+  auto values = rmm::device_uvector<size_type>(num_rows, stream, mr);
```
Comment (PointKernel) [inline cpp/src/stream_compaction/distinct.cu:136]:
```suggestion
  auto keys   = rmm::device_uvector<size_type>(num_rows, stream);
  auto values = rmm::device_uvector<size_type>(num_rows, stream);
```
only the return value needs to be created via the given `mr`

**Comment by PointKernel** [comment]
@srinivasyadav18  can you please share the performance comparisons before and after this PR?

**Comment by GregoryKimball** [comment]
Based on this [gist](https://gist.github.com/srinivasyadav18/a353458f09d790c83e1c18ef04044814), here are the most important performance signals:
image

I'm comparing the 1B row case, removing `any` because it's not impacted by this PR, and removing `last` because it is identical to `first`. 

* drop at 1B for `keep=first`
* drop at medium-low cardinality (1-100K)
* improvement at low cardinality (100) for I32
* improvement at 1-100M

Overall, I believe this performance signature is a wash. Perhaps the shifting throughputs are due to cache locality differences. It's worth profiling the `1B rows, 1B cardinality, keep first` case and the `1B rows, 100K cardinality, keep first` case.

**Comment by srinivasyadav18** [comment]
Thanks @GregoryKimball for providing the plot and important performance signals.

I have done some profiling on 1 Billion keys and 1 Billion cardinality, and this what I found out.

Overall distint algorithm has 2 major steps
-  distinct_indices 
     - initalize data strucutre
     - perform the reductions (reduce_by_row)
     - return the output_indices (all the distinct indices based on keep option)
- gather

Here is what the base branch-24.10 [old] numbers look like 

-  distinct_indices 
     - initalize data strucutre (cuco::static_set, reduction_results device_uvector) **TIME = 5ms**
     - perform the reductions (reduce_by_row) **TIME = 65.589ms**
     - return the output_indices (all the distinct indices based on keep option) **TIME = 4ms**
- gather **##TIME = 15ms##**  

Below is the profiling of base branch-24.10 (selected region in green shade shows gather runtime)

base


Here is the new implementation **insert_or_apply** numbers look like
-  distinct_indices 
     - initalize data strucutre                                                                                    **TIME = 7.8ms**
     - perform the reductions (insert_or_apply) (thrust::for_each with set_ref)        **TIME = 55.7ms**
     - return the output_indices (retrieve_all)                                                           **TIME = 11.8ms**
- gather                                                                                                                   **##TIME = 53.364##**

Below is the profiling with insert_or_apply  (selected region in green shade shows gather runtime)
insert_or_apply_gather

In summary, the gather algorithm is causing the main bottleneck.
It looks like the gather algorithm is efficient only when the row indices are already sorted (this makes sense because there is less data movement).

I tested the **insert_or_apply** implementtation to sort the array before returning from the distinct_algorithm, and performance improved signifcantly as now gather takes very less time. 

Below is the profiling with insert_or_apply with sorting the distinct indices:  (selected region in green shade shows gather runtime) (red shows the sorting overhead)

sort_insert_or_apply_gather

**Comment by srinivasyadav18** [comment]
In summary, the branch-24.10 implementation is already efficient as it performs reduction and also maintains the ordering for reduction values (as they are row-indices). 

`insert-or-apply` kernel improves performance more than 15%, but due to unordered indices as the output, the gather algorithm which does post-processing regresses the performance more than 50%, hence causing the overall performance regression in the pipeline. 

Sorting helps the situation in large input cases, but causes regression in low input cases (see here in [gist ](https://gist.githubusercontent.com/srinivasyadav18/2c32833580f00aafd4d650f037096d4d/raw/fec500995c311c2a733dcb377ed7413705902d2a/h100_pcie_sort_ioa.txt)for numbers).

**Comment by GregoryKimball** [comment]
Thank you @srinivasyadav18 for studying the 1B cardinality case. The gist you posted shows the benefit.
image

It would be worth a profile and check of the 100K cardinality case, but I don't think the moderate cardinality performance is a blocker

**Comment by ttnghia** [comment]
Instead of sorting, how about using scatter-gather approach?
 * Initialize a gather map with all `0`
 * Scatter the output indices into the gather map, marking the position of these indices as `1`
 * Gather using that gather map.

File: cpp/src/stream_compaction/distinct.cu [hunk 4]
```diff
@@ -96,17 +99,39 @@ rmm::device_uvector<size_type> distinct_indices(table_view const& input,
...
+                                          {},
+                                          {},
+                                          cudf::detail::cuco_allocator{stream},
+                                          stream.value()};
+
+      auto const iter = thrust::counting_iterator<cudf::size_type>{0};
+      set.insert_async(iter, iter + num_rows, stream.value());
+      auto output_indices   = rmm::device_uvector<size_type>(num_rows, stream, mr);
+      auto const output_end = set.retrieve_all(output_indices.begin(), stream.value());
+      output_indices.resize(thrust::distance(output_indices.begin(), output_end), stream);
```
Comment (ttnghia) [inline cpp/src/stream_compaction/distinct.cu:120]: Can we know the output size before calling to `retrieve_all`?

File: cpp/src/stream_compaction/distinct_helpers.cu [hunk 5]
```diff
@@ -22,118 +22,105 @@
...
-      output_indices.begin(),
-      cuda::proclaim_return_type<bool>([init_value = reduction_init_value(keep)] __device__(
-                                         auto const idx) { return idx != init_value; }));
-  }();
+  auto const map_end = thrust::copy_if(
+    rmm::exec_policy_nosync(stream),
+    output_iter,
+    output_iter + num_distinct_keys,
+    output_indices.begin(),
+    cuda::proclaim_return_type<bool>([] __device__(auto const idx) { return idx != -1; }));
```
Comment (ttnghia) [inline cpp/src/stream_compaction/distinct_helpers.cu:83]: Please add back the comment.  Although I wrote the original code, now I hardly understand what it is doing without the comment.

File: cpp/src/stream_compaction/distinct_helpers.hpp [hunk 6]
```diff
@@ -47,6 +49,31 @@ auto constexpr reduction_init_value(duplicate_keep_option keep)
   }
 }
 
+struct plus_op {
+  template <cuda::thread_scope Scope>
+  __device__ void operator()(cuda::atomic_ref<size_type, Scope> ref, size_type const val)
```
Comment (ttnghia) [inline cpp/src/stream_compaction/distinct_helpers.hpp:54]: Please do not put `__device__` code in `hpp` file.

File: cpp/src/stream_compaction/distinct_helpers.hpp [hunk 7]
```diff
@@ -23,11 +23,13 @@
 #include <rmm/device_uvector.hpp>
 #include <rmm/resource_ref.hpp>
 
+#include <cuco/static_map.cuh>
```
Comment (ttnghia) [inline cpp/src/stream_compaction/distinct_helpers.hpp:26]: Now with including the `cuco` header, this file should be renamed into `_helper.cuh`. Otherwise please move the these header and their relevant code into `_helper.cuh` and keep this file containing host-only code.

**Comment by GregoryKimball** [comment]
We hope that https://github.com/rapidsai/cudf/pull/15700 will improve the overall performance outlook of `static_map` in `distinct`

**Comment by PointKernel** [comment]
Update: waiting for #15700 to merge to determine if the current PR can improve performance after resolving the register issues.

**Comment by GregoryKimball** [comment]
We could re-benchmark this feature using a primitive row operator as in https://github.com/rapidsai/cudf/pull/18896

**Comment by vyasr** [comment]
I'm going to close this given how out of date the implementation is. If there's still interest of porting this over to the latest version, that can be done later, but I assume it will need some rebuilding from scratch.
